internal/catalogmetadata/cache/cache.go

```diff
@@ -0,0 +1,152 @@
+package cache
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"os"
+	"path/filepath"
+	"sync"
+
+	catalogd "github.com/operator-framework/catalogd/api/core/v1alpha1"
+	catalogmetadataclient "github.com/operator-framework/operator-controller/internal/catalogmetadata/client"
+)
+
+// TODO: When a catalogd release containing https://github.com/operator-framework/catalogd/pull/168
+// is made this can be removed in favor of parsing `Catalog.Status` for the
+// appropriate URL for fetching catalog contents.
+const catalogdOnClusterBaseURL = "http://catalogd-catalogserver.catalogd-system.svc"
+const catalogdOnClusterURLTemplate = "%s/catalogs/%s/all.json"
+
+var _ catalogmetadataclient.Cache = &FilesystemCache{}
+
+func NewFilesystemCache(cachePath string, tripper http.RoundTripper) *FilesystemCache {
+	return &FilesystemCache{
+		cachePath:              cachePath,
+		mutex:                  sync.RWMutex{},
+		tripper:                tripper,
+		cacheDataByCatalogName: map[string]cacheData{},
+	}
+}
+
+// cacheData holds information about a catalog
+// other than it's contents that is used for
+// making decisions on when to attempt to refresh
+// the cache.
+type cacheData struct {
+	LastModified string
+	Generation   int64
+	ResolvedRef  string
+}
+
+// FilesystemCache is a cache that
+// uses the local filesystem for caching
+// catalog contents. It will fetch catalog
+// contents if the catalog does not already
+// exist in the cache.
+type FilesystemCache struct {
+	mutex                  sync.RWMutex
+	cachePath              string
+	tripper                http.RoundTripper
+	cacheDataByCatalogName map[string]cacheData
+}
+
+// FetchCatalogContents will fetch the contents for the provided
+// Catalog from the filesystem. If the provided Catalog has not yet
+// been cached, it will make a GET request to the Catalogd HTTP server
+// to get the Catalog contents and cache them. The cache will be updated
+// automatically if a Catalog is noticed to have a different generation
+// and resolved image reference.
+// The Catalog provided to this function is expected to:
+// - Be non-nil
+// - Have a Catalog.Status.Conditions entry with type Unpacked and status True
+// - Have a non-nil Catalog.Status.ResolvedSource.Image
+// This ensures that we are only attempting to fetch catalog contents for Catalog
+// resources that have been successfully reconciled, unpacked, and are being served.
+// These requirements help ensure that we can rely on status conditions to determine
+// when to issue a request to update the cached Catalog contents.
+func (fsc *FilesystemCache) FetchCatalogContents(ctx context.Context, catalog *catalogd.Catalog) (io.ReadCloser, error) {
+	fsc.mutex.Lock()
```

Comment:
When you have to check whether or not the cache needs updating, be careful not to acquire the write lock too early, as it's exclusive. We can infer from our system architecture that reads will outnumber writes by a large clip, so a less intrusive implementation would look like:

```
read-lock
check preconditions
read-unlock

if preconditions met to write new entry
fetch new data
acquire write lock
check preconditions
ensure nobody raced and updated before us
if preconditions are still in the same place we thought they were at the time we had the read lock
write the data
write unlock
```

Reply:
Updated this in https://github.com/operator-framework/operator-controller/pull/411/commits/fe924c2c2f9c686d3720bb930d42f01590409df1 - thanks for pairing with me on this to help me understand this better!